pyron-engine/src/main/scala/com/cloudentity/pyron/apigroup/ApiGroupConflicts.scala

```diff
@@ -6,24 +6,25 @@ object ApiGroupConflicts {
 
   case class Conflict(pathA: List[String], pathB: List[String])
 
-  def findConflicts(validGroups: List[ValidResult[ApiGroupConfUnresolved]]): List[Either[Conflict, ValidResult[ApiGroupConfUnresolved]]] =
-    validGroups.map { validGroup =>
-      val otherConflictedGroupOpt =
-        validGroups.find { otherGroup =>
-          validGroup.path != otherGroup.path && isConflicted(validGroup.value.matchCriteria, otherGroup.value.matchCriteria)
-        }
-
-      otherConflictedGroupOpt match {
-        case Some(otherConflictedGroup) => Left(Conflict(validGroup.path, otherConflictedGroup.path))
-        case None                       => Right(validGroup)
-      }
+  def findConflicts(validGroups: List[ValidResult[ApiGroupConfUnresolved]]): List[Either[Conflict, ValidResult[ApiGroupConfUnresolved]]] = {
+    val validAndConflicted = for {
+      (validGroup, idx) <- validGroups.zipWithIndex
+    } yield (validGroup, validGroups.drop(idx + 1).find { otherGroup =>
```

Comment:
ok, there's this thing, if some rule comes earlier and is more specific, it should not conflict with more general rule which comes later. this is how nginx works, and it's hard to replace it without doing sth equivalent. For example it allows to catch remaining stuff, after we capture all special cases. So now we check forward for more specific rules conflicting, but not backwards. It is more relaxed than it was, so it won't brake existing checks.

Reply:
The reason for having ApiGroups is to be able to separate APIs in non-overlapping groups, i.e. to be sure that an API from one ApiGroup is not overwritten by an API from another ApiGroup (e.g. to allow multi-tenancy). Relaxing `findConflicts` breaks this guarantee. We should not change it and instead find another way, e.g. change how UI works.

Reply:
I'm not sure if we can say, it's overlapping, if in fact no other group matched. For example We may have apiGroup "login" for path: "/login" and apiGroup "root" for path "/". If the request is "/" -> "root" is the only apiGroup that matches. There isn't really overlap in such case. On the other hand "/login/stuff" could be matched by both, although "/login" is more specific and defined earlier. I need to serve the case of matching "/" when other, more specific paths did not match.

I agree rule name order which I used is probably too lousy. Instead we could order rules by path segment count, and do not allow conflict between rules of same segment count. We would try to match starting with the most specific (max segment count), if no match found, try shorter, more general and so on. Wouldn't change anything for existing configurations, and would solve the problem. I don't think there's really a conflict between general and more specific definitions, the more specific would simply win. So I think this would be the best approach.

Other solution could be to allow for "default" rule or fallback "rules": [...] after the "apiGroups" definition. This could be in addition to the previous approach I described, or without it. Currently it's not allowed to have both "apiGroups" and "rules" but it could be easily changed in `private def buildApiGroupsOrFallbackToRules` in ApiGroupStore. Anyways, we need some way to catch the "reminder" or more general case after more specific did not capture the path.

I did it by relaxing the "conflict" criteria, but it could be done differently. So perhaps one of the solutions I state above would be acceptable? If not, then please give your own suggestion.

Reply:
As discussed, api-groups allow separation of 2 teams using the same Pyron instance without getting in their way. Let's figure out how to address '/' issue.